src/main/java/frc/robot/util/BetterPID.java

```diff
@@ -0,0 +1,464 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.util;
+
+import edu.wpi.first.math.MathSharedStore;
+import edu.wpi.first.math.MathUsageId;
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.util.sendable.Sendable;
+import edu.wpi.first.util.sendable.SendableBuilder;
+import edu.wpi.first.util.sendable.SendableRegistry;
+
+/** Implements a PID control loop. */
+public class BetterPID implements Sendable, AutoCloseable {
```

Comment:
Add a link to the original from which this copied: a) to give them credit, and b) so that people can go see the original.

What is changed from the original? Why are we using this in this case? Should we use it everywhere, or only in specific cases? Answer these questions here, clearly, in the comments on this class.

Reply:
If you look in the calculate method of the PID (around [here] (https://github.com/Stealth-Robotics/Stealth2024/blob/a34c80b4ec30f1e477e1f6d66967a53b60c226e5/src/main/java/frc/robot/util/BetterPID.java#L430-L432) you'll see some conditional integration. I worked with Alex on this, and basically what the condition is checking is if the input is saturated (that's what `calc > 0.9` is doing.) Secondly, it also checks if the integral buildup is going the wrong way, by checking the signs of the error and calculation. If both these are true, it means we have integral buildup and we won't integrate.

Reply:
Ah, I see. I got the diff backwards because of the way I tried to look at it. That sounds reasonable, and the code looks reasonable.

Honestly, I'd put your explanation above in a comment in the code! It's good.

The only other question I have now, and others will have later, is should you use this all the time, or only in special circumstances? Leave a note about that in there too.

Reply:
yea, i could just overload the constructor in it. Doing that now.

Reply:
Oh, I'm not asking for you to change the code to make options. I'm asking you to guide others with **_why_** to use this or not!

Reply:
Let me give you some possible suggestions. I don't know which of these is correct, though, you need to tell us. I'd assume none of them are actually correct, but perhaps one or two could be close.

- I made this pid for fun, 'cause integrals sound cool. Use it or not, idk.
- This PID adds more anit-integral windup logic to help with systems which are slow to reach their setpoint. In these cases, you end up with a large accumulated error, and a bunch of oscillation before the system finally settles. An arm with a low max velocity might be a good use for this, or a flywheel with a bunch of mass that is slow to spin up. For the 2024 bot, we used it for XXX subsystem. If you use it, you still want to tune iZone and the min/max integral.
- This PID adds more anti-integral windup logic that is widely useful. Every system we have in 2024 will benefit from it, and you should prefer this over the default PID in WPILib whether you have an I term or not!